rust/gateway/src/main.rs

```diff
@@ -98,39 +100,48 @@ async fn get_firezone_id(env_id: Option<String>) -> Result<String> {
 }
 
 async fn run(login: LoginUrl, private_key: StaticSecret) -> Result<Infallible> {
-    let mut tunnel = GatewayTunnel::new(private_key, Sockets::new(), CallbackHandler)?;
-
-    let (portal, init) = phoenix_channel::init::<_, InitGateway, _, _>(
+    let tunnel = GatewayTunnel::new(private_key, Sockets::new(), CallbackHandler)?;
+    let portal = PhoenixChannel::connect(
         Secret::new(login),
         get_user_agent(None, env!("CARGO_PKG_VERSION")),
         PHOENIX_TOPIC,
         (),
         ExponentialBackoffBuilder::default()
             .with_max_elapsed_time(None)
             .build(),
-    )
-    .await??;
-
-    tunnel
-        .set_interface(&init.interface)
-        .context("Failed to set interface")?;
-    let mut tun_device = connlib_shared::tun_device_manager::TunDeviceManager::new()?;
-    tun_device
-        .set_ips(init.interface.ipv4, init.interface.ipv6)
-        .await?;
-    tun_device
-        .set_routes(
-            vec![Cidrv4::from(PEERS_IPV4.parse::<Ipv4Network>().unwrap())],
-            vec![Cidrv6::from(PEERS_IPV6.parse::<Ipv6Network>().unwrap())],
-        )
-        .await?;
-    tunnel.update_relays(HashSet::default(), init.relays);
-
-    let mut eventloop = Eventloop::new(tunnel, portal);
-
-    future::poll_fn(|cx| eventloop.poll(cx))
-        .await
-        .context("Eventloop failed")?;
+    );
+
+    let (sender, mut receiver) = mpsc::channel::<Interface>(10);
+
+    let mut tun_device = TunDeviceManager::new()?;
+
+    let update_device_task = async move {
+        while let Some(next_interface) = receiver.next().await {
+            if let Err(e) = tun_device
+                .set_ips(next_interface.ipv4, next_interface.ipv6)
+                .await
+            {
+                tracing::warn!("Failed to set interface: {e:#}");
+            }
+
+            if let Err(e) = tun_device
+                .set_routes(
+                    vec![Cidrv4::from(PEERS_IPV4.parse::<Ipv4Network>().unwrap())],
+                    vec![Cidrv6::from(PEERS_IPV6.parse::<Ipv6Network>().unwrap())],
+                )
+                .await
+            {
+                tracing::warn!("Failed to set routes: {e:#}");
+            };
+        }
+    };
+
+    let mut eventloop = Eventloop::new(tunnel, portal, sender);
+    let eventloop_task = future::poll_fn(move |cx| eventloop.poll(cx));
+
+    let ((), result) = futures::join!(update_device_task, eventloop_task);
```

Comment:
do we need to join? Should we just `spawn` the device task and then await the eventloop?

Reply:
It doesn't matter much here but I try to avoid spawn as much as possible to not run into resource leaks.

Reply:
I think structured concurrency is the much better building block.

Reply:
If you haven't yet, I can highly recommend reading https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/.